python/mxnet/gluon/data/sampler.py

```diff
@@ -136,3 +138,74 @@ def __len__(self):
         raise ValueError(
             "last_batch must be one of 'keep', 'discard', or 'rollover', " \
             "but got %s"%self._last_batch)
+
+
+class AliasMethodSampler(object):
+    """ The Alias Method: Efficient Sampling with Many Discrete Outcomes.
+    Can be use in NCELoss.
+
+    Parameters
+    ----------
+    K : int
+        Number of events.
+    probs : array
+        Probability of each events, corresponds to K.
+
+    References
+    -----------
+        https://hips.seas.harvard.edu/blog/2013/03/03/the-alias-method-efficient-sampling-with-many-discrete-outcomes/
+    """
+    def __init__(self, K, probs):
+        if K != len(probs):
+            raise ValueError("K should be equal to len(probs). K:%d, len(probs):%d" % (K, len(probs)))
+        self.K = K
+        self.prob = nd.zeros(K)
+        self.alias = nd.zeros(K, dtype='int32')
+
+        # Sort the data into the outcomes with probabilities
+        # that are larger and smaller than 1/K.
+        smaller = []
+        larger = []
+        for kk, prob in enumerate(probs):
+            self.prob[kk] = K*prob
+            if self.prob[kk] < 1.0:
```

Comment:
Is `asscalar()` missing here? Is using NDArray actually faster thank numpy array? This comparison converts a ndarray to a scalar. In my experience using a python list is fastest, followed by numpy array, followed by ndarray. We can convert the list to mxnet ndarray after the O(N) preprocessing. 
You can try a distribution with 1M classes to compare the speed. 
@zihaolucky  @szha

Reply:
#9209 FYI, a relevant issue here.

Reply:
I would compare the speed later and report here.

Reply:
@eric-haibin-lin The test code `speed` in my server, the Numpy is 100x faster than NDArray.